hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/DefaultCertificateClient.java

```diff
@@ -1095,4 +1103,314 @@ public synchronized void close() throws IOException {
+      throw new CertificateException("Error while signing and storing new" +
+          " certificates.", e, RENEW_ERROR);
+    }
+
+    // switch Key and Certs directory on disk
+    File currentKeyDir = new File(
+        securityConfig.getKeyLocation(component).toString());
+    File currentCertDir = new File(
+        securityConfig.getCertificateLocation(component).toString());
+    File backupKeyDir = new File(
+        securityConfig.getKeyLocation(component).toString() +
+            HDDS_BACKUP_KEY_CERT_DIR_NAME_SUFFIX);
+    File backupCertDir = new File(
+        securityConfig.getCertificateLocation(component).toString() +
+            HDDS_BACKUP_KEY_CERT_DIR_NAME_SUFFIX);
+
+    if (!currentKeyDir.renameTo(backupKeyDir)) {
+      // Cannot rename current key dir to the backup dir
+      throw new CertificateException("Failed to rename " +
+          currentKeyDir.getAbsolutePath() +
+          " to " + backupKeyDir.getAbsolutePath() + " during " +
+          "certificate renew.", RENEW_ERROR);
+    }
+    if (!currentCertDir.renameTo(backupCertDir)) {
+      // Cannot rename current cert dir to the backup dir
+      rollbackBackupDir(currentKeyDir, currentCertDir, backupKeyDir,
+          backupCertDir);
+      throw new CertificateException("Failed to rename " +
+          currentCertDir.getAbsolutePath() +
+          " to " + backupCertDir.getAbsolutePath() + " during " +
+          "certificate renew.", RENEW_ERROR);
+    }
+
+    if (!newKeyDir.renameTo(currentKeyDir)) {
+      // Cannot rename new dir as the current dir
+      String msg = "Failed to rename " + newKeyDir.getAbsolutePath() +
+          " to " + currentKeyDir.getAbsolutePath() +
+          " during certificate renew.";
+      // rollback
+      rollbackBackupDir(currentKeyDir, currentCertDir, backupKeyDir,
+          backupCertDir);
+      throw new CertificateException(msg, RENEW_ERROR);
+    }
+
+    if (!newCertDir.renameTo(currentCertDir)) {
+      // Cannot rename new dir as the current dir
+      String msg = "Failed to rename " + newCertDir.getAbsolutePath() +
+          " to " + currentCertDir.getAbsolutePath() +
+          " during certificate renew.";
+      // delete new key directory
+      try {
+        Files.createDirectories(currentKeyDir.toPath());
+      } catch (IOException e) {
+        throw new CertificateException(msg, RENEW_ERROR);
+      }
+      // rollback
+      rollbackBackupDir(currentKeyDir, currentCertDir, backupKeyDir,
+          backupCertDir);
+      throw new CertificateException(msg, RENEW_ERROR);
+    }
+
+    getLogger().info("Successful renew key and certificate." +
+        " New certificate {}.", newCertSerialId);
+    return newCertSerialId;
+  }
+
+  private void rollbackBackupDir(File currentKeyDir, File currentCertDir,
+      File backupKeyDir, File backupCertDir) throws CertificateException {
+    // move backup dir back as current dir
+    if (!currentKeyDir.exists() && backupKeyDir.exists()) {
```

Comment:
This condition seems to be problematic, in the case when we already moved parts of the new key and cert to the current location.
Let's assume we successfully move the newKeyDir to currentKeyDir, and then moving newCertDir to currentCertDir fails, and we call this method. In that case currentKeyDir.exists() will return true, which makes this expression to evaluate to false, and we do not do rollback, however we indeed should, as if we don't, and we report just a RENEW_ERROR, then things go on with the new keys and the old certificates directory as I think.

So I believe we just need to check if the backup directory exists, also I would consider whether we should log the fact that we skipped rollback as the backup directory does not exists.
We might also switch to Files.move and use the REPLACE_EXISTING CopyOption, as with renameTo we would need to check if the target exists and remove first I guess that is why the non-existence check is added here at the first place.

Reply:
> This condition seems to be problematic, in the case when we already moved parts of the new key and cert to the current location. Let's assume we successfully move the newKeyDir to currentKeyDir, and then moving newCertDir to currentCertDir fails, and we call this method. In that case currentKeyDir.exists() will return true, which makes this expression to evaluate to false, and we do not do rollback, however we indeed should, as if we don't, and we report just a RENEW_ERROR, then things go on with the new keys and the old certificates directory as I think.
> 

 currentKeyDir is deleted before it comes to rollbackBackupDir function, as you have suggested that there is no need to move currentKeyDir back t o newKeyDir. so it's deleted before rollbackBackupDir function.  What  rollbackBackupDir does is just as the name indicating, move backupDir back as currentDir.  So the case you mentioned actually will not happen.